### ISSUE TITLE
Noise measurement fails

### ISSUE THREAD

**Issue by SoFiA-Admin** [SoFiA-Admin/SoFiA#85]
When trying to run a simple, absolute threshold finder on a data cube, SoFiA crashes with the following messages:

    --- SoFiA: Measuring cube noise ---
    ~/SoFiA/git/build/lib.linux-x86_64-2.7/sofia/functions.py:33: RuntimeWarning: divide by zero encountered in double_scalars
      bins=np.arange(cubemin,abs(cubemin)/nrbins-1e-12,abs(cubemin)/nrbins)
    Traceback (most recent call last):
      File "~/SoFiA/git/sofia_pipeline.py", line 223, in <module>
        Estimating rms on subcube (x,y,z zoom = 1,1,1) ...
        Estimating rms on subcube sampling every 1 voxels ...
        ... Subcube shape is (86, 167, 198) ...
        globalrms=functions.GetRMS(np_Cube,rmsMode='negative',zoomx=1,zoomy=1,zoomz=1,verbose=True,sample=sampleRms)
      File "~/SoFiA/git/build/lib.linux-x86_64-2.7/sofia/functions.py", line 33, in GetRMS
        bins=np.arange(cubemin,abs(cubemin)/nrbins-1e-12,abs(cubemin)/nrbins)
    ValueError: Maximum allowed size exceeded
    Pipeline failed with exit code 1.

The only modules enabled are: threshold finder with an absolute threshold of 0.05, merging of detections, and output of ASCII and XML catalogues. Most of the cube is zero, i.e. there is no noise in it, so I wonder whether that causes the noise measurement to crash (rather than report the correct noise level of 0).

**Comment by ngiese** [comment]
The problem is probably related to the zeros in the cube. The GetRMS module adjusts the bin size of the histogram of which the negative part is fit according to several criterions. If all values are zero, these criterions are never met. We should probably fix it and for your data cube, you could use another way of estimating the noise (std or mad). Maybe the best way to solve this is falling back to alternative methods if one fails.

**Comment by SoFiA-Admin** [comment]
Thanks for your suggestions. Where do I change the noise method, and why is the noise being measured at all even though I use an absolute threshold?

**Comment by ngiese** [comment]
The noise is not measured in the threshold module if you use an absolute threshold. It is measured somewhere at the beginning of the pipeline. It is later used to normalize the fluxes of the sources (snr_min/max/sum). If the noise level is zero, however, this will fail. If you want to change this behaviour, you can do it in sofia_pipeline.py, lines 223 and 292-294.

**Comment by SoFiA-Admin** [comment]
Thanks for locating the issue. The “negative” method is actually hard-coded into the pipeline at the moment. I think this will need to be handled differently in the future. At the same time, we need to carry out checks after every single step to ensure that the result of any calculation is actually valid and doesn’t crash the pipeline. With the current implementation it is not possible to run an external de-noising filter on a cube and then feed the output into SoFiA for thresholding, merging and parametrisation (which is what I tried to do).

**Comment by SoFiA-Admin** [comment]
By the way, I just noticed that we can't use the 2D-1D wavelet filter in SoFiA any longer, because it too produces noise-free reconstructed cubes (which is its intended purpose), and the mandatory noise measurement in the pipeline will therefore crash again.

**Comment by paoloserra** [comment]
This has now been fixed. The cause of the crash was the attempt to measure the global rms using the negative histogram method on a cube with no noise.

We realised that this global rms value is really only used during the reliability calculation, which in turn should not be attempted for a cube with no noise. Therefore, it makes sense to only measure the global rms within the reliability module.

For a (filtered) input cube with no noise the reliability calculation should be switched off, and therefore no global rms measurement will be attempted.

For more details see https://github.com/SoFiA-Admin/SoFiA/commit/2d712370a746be9f924f3e5bb5eec77469cc329b .
